### ISSUE TITLE
V2020.2 : Incorrect 'Today' readings for 'Energy read: Computed' General Electric sensors with negative instant value

### ISSUE THREAD

**Issue by mvdklip** [domoticz/domoticz#4736]
When updating a General Electric (kWh) sensor which has been set to 'Energy read: Computed' **with a negative instant value** (wattage), the 'Today' value displayed in the web interface is correctly negative but not the correct value. The Today value seems to 'jump' back to 0 regularly.

I have created a test plugin which sends the same wattage to two sensors, one being a negative value and the other a positive. It updates every 5 seconds and lets Domoticz compute the counter. The counters seems to be fine if you look at the logging after a bit of time:

    2021-03-19 14:04:41.756 (testit - Negative wattage) Updating device from 0:'-3600;-1278.0' to have values 0:'-3600;0'.
    2021-03-19 14:04:41.765 (testit - Positive wattage) Updating device from 0:'3600;1278.0' to have values 0:'3600;0'.

Both counters are the same except for the sign. However if you look at the web interface something's not right:

![Negative wattage](https://user-images.githubusercontent.com/200318/111785208-98940a00-88bc-11eb-8c21-8d70cec1279f.png)

The Today value for the negative wattage is way off. If you keep looking at this value you can see it jump sometimes.

Tested on the latest stable version (2020.2). I am unable to test this on the latest beta right now.

**Update 1: this problem has been confused with another 'toggling' problem introduced by a fix for #4733. Although that is not the same problem as described above a lot of the comments below are about this toggling problem.**

**Update 2: this has been tested on beta build 13111 and the problem doesn't seem to manifest itself there so it seems limited to the latest stable.**

**Comment by mvdklip** [comment]
"""
	<plugin key="testit" name="testit" author="mvdklip" version="1.0.0">
		<description>
			<h2>Test Plugin</h2><br/>
		</description>
		<params>
			<param field="Mode3" label="Query interval" width="75px" required="true">
				<options>
					<option label="5 sec" value="1" default="true"/>
					<option label="15 sec" value="3"/>
					<option label="30 sec" value="6"/>
					<option label="1 min" value="12"/>
					<option label="3 min" value="36"/>
					<option label="5 min" value="60"/>
					<option label="10 min" value="120"/>
				</options>
			</param>
		</params>
	</plugin>
	"""

	import Domoticz


	class BasePlugin:
		enabled = False
		lastPolled = 0

		def __init__(self):
			return

		def onStart(self):
			Domoticz.Debug("onStart called")
			Domoticz.Debugging(1)

			if len(Devices) < 1:
				Domoticz.Device(Name="Negative wattage", Unit=1, TypeName='kWh', Options={'EnergyMeterMode':'1'}).Create()
			if len(Devices) < 2:
				Domoticz.Device(Name="Positive wattage", Unit=2, TypeName='kWh', Options={'EnergyMeterMode':'1'}).Create()

			DumpConfigToLog()

			Domoticz.Heartbeat(5)

		def onStop(self):
			Domoticz.Debug("onStop called")

		def onHeartbeat(self):
			Domoticz.Debug("onHeartbeat called %d" % self.lastPolled)

			if self.lastPolled == 0:
				Devices[1].Update(nValue=0, sValue="-3600;0")
				Devices[2].Update(nValue=0, sValue="3600;0")

			self.lastPolled += 1
			self.lastPolled %= int(Parameters["Mode3"])


	global _plugin
	_plugin = BasePlugin()


	def onStart():
		global _plugin
		_plugin.onStart()


	def onStop():
		global _plugin
		_plugin.onStop()


	def onHeartbeat():
		global _plugin
		_plugin.onHeartbeat()


	# Generic helper functions
	def DumpConfigToLog():
		for x in Parameters:
			if Parameters[x] != "":
				Domoticz.Debug("'" + x + "':'" + str(Parameters[x]) + "'")
		Domoticz.Debug("Device count: " + str(len(Devices)))
		for x in Devices:
			Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
			Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
			Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
			Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
			Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
			Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
		return

**Comment by FireWizard52** [comment]
I can confirm related strange behaviour.
See for the details: https://www.domoticz.com/forum/viewtopic.php?f=6&t=35839

**Comment by kovainfo** [comment]
all kWh sensors stopped working

**Comment by rwaaren** [comment]
> all kWh sensors stopped working

This statement is nor true nor does it bring us it any closer to finding a cause for this issue.
Please add some details like your domoticz version, settings of your device and how they are updated.


I am on build 13091 and  my kWh sensors updated by dzVents and set as 'from device' are working as before. 

![image](https://user-images.githubusercontent.com/24454798/111835101-cf861200-88f4-11eb-96f8-d391fb70d899.png)


![image](https://user-images.githubusercontent.com/24454798/111834244-99945e00-88f3-11eb-8f41-a59e0ad64db2.png)

![image](https://user-images.githubusercontent.com/24454798/111834440-e710cb00-88f3-11eb-97ee-a954acbbc318.png)

**Comment by FireWizard52** [comment]
Hello @rwaaren 

I can imagine, that it is working in your case, as you use "From device".
I' m on 13091 as well, but all my devices with a configuration "Computed"  are failing. The devices with "From device" are okay.

The issue is introduced in 13088, yesterday.
See also: https://github.com/domoticz/domoticz/issues/4733

And also: https://www.domoticz.com/forum/viewtopic.php?f=6&t=35839.

For me, it is pretty clear, but I am not able to reverse the change.

**Comment by rwaaren** [comment]
@FireWizard52, I already was aware that the issue was with the devices set to computed. The reason that I commented was to prevent someone from investigating what I already noticed: that devices set to "From device" where not hampered by this issue.
Without my comment developers looking into this might have been set on the wrong foot by the statement of @kovainfo that **All** KWh devices stopped working.

**Comment by kovainfo** [comment]
Ok, all MY kWh devices stopped working 🤷🏻‍♂️

now I have rolled back to the previous version of Domoticz, which worked successfully for me (
Build Hash: 066f3f261-modified
Build Date: 2021-03-08 18:27:50
)

The error manifested itself like this:
At first, the "Today" counter showed a negative value, and then the values simply stopped being displayed.

I use only two types of kWh devices: a custom device that is updated through Domoticz API and Xiaomi kWh devices from sockets, etc.

And, yes, my devices havе the option: "Computed"

**Comment by mvdklip** [comment]
Please comment in #4733 if your meters stopped working / started toggling. That behaviour seems to be introduced by a supposed fix for #4733.

Although I'm the reporter of both issues, they are not related. This issue is about something else which manifests itself in the latest stable (2020.2.11995).

**Comment by kovainfo** [comment]
Domoticz 2020.2 (build 13058)
<img width="810" alt="Снимок экрана 2021-03-20 в 14 02 28" src="https://user-images.githubusercontent.com/36986231/111867637-6c44c000-8986-11eb-8440-696e03948634.png">


but last beta:
<img width="810" alt="Снимок экрана 2021-03-20 в 14 24 30" src="https://user-images.githubusercontent.com/36986231/111867631-64851b80-8986-11eb-9e7e-e3a2b69d2fe5.png">

**Comment by devrosx** [comment]
same problem here with latest beta
Return type works but usage type is empty... (im using computed energy read)
for moment it works but then is somehow erased...

**Comment by gizmocuz** [comment]
There was a issue in the computed energy routine. The crash had been solved, but introduced another issue.
A new beta has been made and the problem should be solved now.
Please close the issue if that's the case ;)

**Comment by FireWizard52** [comment]
I loaded the new beta (version 13092), but another issue has been introduced.
If you set the type to "Return" and "Energy read" has been set to "Computed" the devices switch back to default, but the selector field (Usage/Return) is empty.

To reproduce:
1. Set type to "Return"
2. Set "Energy read" to "Computed"

[Added]  Also the values are toggling still. So it is not solved, but it is getting worse.

**Comment by devrosx** [comment]
Yes today count is still empty and values are still toggling with latest build....

**Comment by gizmocuz** [comment]
Things are getting confused now... Two related issues does also not help.
If this is a issue that was introduced by 2020.2 stable, yeah it won't be solved by any beta version until someone makes a PR to solve it

**Comment by kovainfo** [comment]
<img width="803" alt="Снимок экрана 2021-03-21 в 13 20 48" src="https://user-images.githubusercontent.com/36986231/111901311-61526400-8a48-11eb-9b44-34a49b947872.png">

:(((((

**Comment by kovainfo** [comment]
build 13058 works fine for me, but now I accidentally erased it and now I don't know what to do (

**Comment by FireWizard52** [comment]
@gizmocuz 

I already said so, that things are getting worse, you call it confused.
It has nothing to do with stable. In my opinion it has been introduced in 13088. I did not install that version, but noticed the issue for the first time in 13090 and also 13091.
13092 did not solve it, but introduced, beside the issue with the toggle of the value  (see screenshot of @kovainfo), also the issue with the setting of Usage/Return. 

Currenly I have set the "Energy read" to "From device" and that gives at least the instant value, but of course the accumulated value is missing.

This gives a stable situation.

**Comment by kovainfo** [comment]
@gizmocuz 
I installed the latest working beta a week ago (March 13-14). for me this version was stable (build 13058).

I think the changes that were made to the kWh device last week and caused the indicated problems




<img width="737" alt="Снимок экрана 2021-03-21 в 13 58 50" src="https://user-images.githubusercontent.com/36986231/111902263-aa58e700-8a4d-11eb-9742-cdf2e1f21db2.png">

**Comment by FireWizard52** [comment]
The problem is still there.

To reproduce:

Set the configuration as follows:

![Screenshot_Energy Computed1](https://user-images.githubusercontent.com/46283323/111919375-f164cd80-8a89-11eb-9470-9e8f4997f78f.png)

With the Type as return.
Close this window
Inject a value (e.g. by MQTT/Node Red)
The icon changes to standard (as for Usage) and the selector field is empty as on the picture.
Configure "Return"
Inject once again it is okay, but change another the first one changes again.

Change the second one, the first one toggles again.

It looks very random, but there is a relationship.

So still not possible to use "Computed" and "Return"

**Comment by gizmocuz** [comment]
@FireWizard52 , there are two issues that are maybe related to each other.
The other issue (first created) was about a crash because of a empty initial sValue.
This crash should hopefully be fixed now
Confusing because the author of that thread (mvdklip) is also reporting here.

I will now try to replicate the issue (just using normal JSON to update the meter)

**Comment by gizmocuz** [comment]
I've added a general kwh sensor, set the type to return and energy read=computer:

![image](https://user-images.githubusercontent.com/2265299/111951822-741a7680-8ae4-11eb-931b-a7e55c3f6cfd.png)

Using the following JSON call to update the sensor

  http://127.0.0.1:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=722;3556

![image](https://user-images.githubusercontent.com/2265299/111951999-b774e500-8ae4-11eb-88d1-1f64b45c92df.png)

Every time I update the widget updates and does not toggle, or flip, or changes it's type

**Comment by gizmocuz** [comment]
![image](https://user-images.githubusercontent.com/2265299/111952063-d5dae080-8ae4-11eb-98a6-8b3bedeb73b5.png)

**Comment by devrosx** [comment]
on my side its still toggling im using MQQT like this for example

2021-03-22 08:21:39.922 MQTT: Topic: domoticz/in, Message: {"idx":940,"command":"udevice","svalue":"199;0"}

**Comment by gizmocuz** [comment]
Maybe you/we could test with the JSON API call first ? I am not familiar with this type of device, maybe it is because behind the semicolon there is a zero (POWER;0) ... 
I do some more testing as well

**Comment by gizmocuz** [comment]
It seems to work perfectly with

http://127.0.0.1:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=1625;0

Could you also test with using the JSON call to update the sensor ?

If this keeps working normally, we know the internal Domoticz code is good, and the issue might be related to MQTT

**Comment by mvdklip** [comment]
> @FireWizard52 , there are two issues that are maybe related to each other.
> The other issue (first created) was about a crash because of a empty initial sValue.
> This crash should hopefully be fixed now
> Confusing because the author of that thread (mvdklip) is also reporting here.

I am sorry for the confusion but the two issues are totally unrelated. Unfortunately this issue was 'hijacked' by people reporting the side effects of the initial fix for #4733.

What shall I do? Close this issue and open a new one for this problem? Because, as said, this is a *different* issue with negative wattages.

**Comment by gizmocuz** [comment]
@mvdklip , Best to check if the other issue is now solved and is yes close that issue

Let's keep this issue open for this problem

**Comment by FireWizard52** [comment]
Hello @gizmocuz 

I used to send the Solar data with MQTT to Domoticz.
As you indicated, that you don't have any problems with http JSON call, I rebuilt my Node Red "Function" node in order to send http instead of mqtt. It doesn't make any difference.

The data is toggling.

See:
![Screenshot_Problem Solar-return-computed1](https://user-images.githubusercontent.com/46283323/112012789-c5008e00-8b29-11eb-8d4f-1b00b663d74a.png)

As you can see, some has the "Usage" icon and some still has the "Return" icon. In case of the "Usage" icon, the selector field is empty.  The "Today" counter is not counting at all, probably because it is reset, when it is empty. Is it reset?

In order to avoid any coincidental interference with other programs, I created a new sensor on my test machine and sent every 2 seconds a JSON message, copied from the one you provided and changed only the IDX,.

The result is exactly the same. It toggles every 2 seconds.

![Screenshot_Problem Solar-return-computed2](https://user-images.githubusercontent.com/46283323/112014185-f29a0700-8b2a-11eb-808f-dff46daa83d7.png)

![Screenshot_Problem Solar-return-computed3](https://user-images.githubusercontent.com/46283323/112014223-f9287e80-8b2a-11eb-8bce-3c2740f52052.png)

Conclusion so far: It doesn't make any difference, whether I use MQTT or HTTP.
Problem start, as soon as I use "Energy read" setting as "Computed".
Problem started after upgrade from 13084 to 13090 and in my opinion it is caused by changes introduced in 13088.

I do not know, what to check/test further.If you need more info, let me know.

Regards

**Comment by gizmocuz** [comment]
@FireWizard52 , could you take a step back please and
- Add a dummy hardware if you don't have one already
- Add a dummy kwh meter (instant+counter)
- Edit the new sensor, set it to return and computed
- Use a browser to send the JSON calls to update the device (like my URL above ... with a different idx value)

How does that workout ?

**Comment by FireWizard52** [comment]
I did one more test.

The toggle behavior is independent of the  value in "svalue=1625;0". I doesn't matter what the figure after 1625 is. 0 is okay, but any other figure shows the same. However if you omit the 0, the screen stops to toggle and shows "no value", as shown in the last picture. Even if you use a minus value it continue to toggle but the "Today" shows a minus value.

Another thing, worthwhile to mention, is that you see the value on the "Today" counter, being similar as the last value in the string.
In my opinion this value should be completely ignored, as the setting for "Energy read" has been set to "Computed" and the value should be based on the first (Power) value (1625), calculated by Domoticz.
Perhaps this indicated a direction.

**Comment by FireWizard52** [comment]
> 
    Add a dummy hardware if you don't have one already
    Add a dummy kwh meter (instant+counter)
    Edit the new sensor, set it to return and computed
    Use a browser to send the JSON calls to update the device (like my URL above ... with a different idx value)

How does that workout ?

I already had "Dummy hardware" and a Dummy kWh meter (instant + counter)
The sensor was already prepared (Return and Computed)

As I previous used a small test flow in Node Red, injecting the time stamp every 2 seconds and sending the same URL, my expectations were not very high. And indeed, it toggles after every transmission.
Except, if I use the browser, I get the confirmation that:

> status | "OK"
> -- | --
> title | "Update Device

Anything else, I can do?

**Comment by gizmocuz** [comment]
@FireWizard52 , it works out that you do not use MQTT or any special way to update the sensor, and have a new test sensor to 'test' against
If you open two browser tabs (one with the view on the test widget), and one where you enter the json call, you should see the widget update
And every 2 seconds does not make sense, try once every 10 seconds/minute

And yes, you need to supply the last value (even if that is zero), the sValue is Instance;Counter as documented in the API

**Comment by FireWizard52** [comment]
@gizmocuz

To be clear.

For the test I do not use MQTT, but for the test I use the JSON call, provided by you, earlier today.

>  it works out that you do not use MQTT or any special way to update the sensor, and have a new test sensor to 'test' against

Yes I do/have a dedicated "test" sensor on a dedicated Raspberry Pi with a dedicated Domoticz, version 13104

> If you open two browser tabs (one with the view on the test widget), and one where you enter the json call, you should see the widget update

I did and saw the toggle.

> And every 2 seconds does not make sense, try once every 10 seconds/minute

I know and agree, For normal use it is useless, however if you want to test something you can configure it to do it faster in order to get the final result earlier. To monitor the widget and to observe, whether it toggles or not, a waiting time of 1 minute is not very smart.

> And yes, you need to supply the last value (even if that is zero), the sValue is Instance;Counter as documented in the API

I know, but in an earlier post, you wrote:

> I am not familiar with this type of device, maybe it is because behind the semicolon there is a zero (POWER;0) ...

That was for me the trigger to play with this value. In my case, it has been zero, all time, but, as said, I noticed that if I insert a non-zero value, this value is displayed in the widget, even if the "Energy read" is set to "Computed" 
This is strange, according to the API description:

> (if you choose as type "Energy read : Computed", this is just a "dummy" counter, not updatable because Domoticz will calculate this internally based on the (previous/current) POWER value)

I would not expected to see that value, as I expect that Domoticz will calculate the value itself based on the received POWER.

Regards

**Comment by gizmocuz** [comment]
Could anyone else also try with my instructions above ? I am not able to reproduce this, making it hard to solve.

I created a bash script that updates the sensor every two seconds, and there are no issues at all.
kWh is calculated, correct usage is displayed on the top right, sensor type stays on return/computed

```
#!/usr/bin/env bash

while :
do
        echo "Press [CTRL+C] to stop.."
        wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=1625;0'
        sleep 2
        wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=225;0'
        sleep 2
        wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=123;0'
        sleep 2
done
```

You need to change the ip address and idx value to the correct one

**Comment by devrosx** [comment]
![Screenshot 2021-03-23 at 10 15 51](https://user-images.githubusercontent.com/13789041/112122730-dc3e8a80-8bc0-11eb-8742-2a26c727aebb.jpg)

https://user-images.githubusercontent.com/13789041/112122754-e2cd0200-8bc0-11eb-9d16-5fd59a7e28a5.mov


> Could anyone else also try with my instructions above ? I am not able to reproduce this, making it hard to solve.
> 
> I created a bash script that updates the sensor every two seconds, and there are no issues at all.
> kWh is calculated, correct usage is displayed on the top right, sensor type stays on return/computed
> 
> ```
> #!/usr/bin/env bash
> 
> while :
> do
>         echo "Press [CTRL+C] to stop.."
>         wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=1625;0'
>         sleep 2
>         wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=225;0'
>         sleep 2
>         wget -q -O /dev/null 'http://192.168.0.123:8080/json.htm?type=command&param=udevice&idx=7828&nvalue=0&svalue=123;0'
>         sleep 2
> done
> ```
> 
> You need to change the ip address and idx value to the correct one

hmm weird you bash script works fine on new created IDX device, but when i change idx in your script to my current device its broken and togling...

**Comment by FireWizard52** [comment]
@gizmocuz 

I tested your provided bash script as well.

I created a new virtual kWh device on the Raspberry Pi with Domoticz 13104.
By default is the virtual sensor configures as "Usage" and "From device"
While the script runs, I reconfigured the virtual sensor to "Return" and "Computed".

Immediately the behavior of the sensor changed to what is shown on the video of @devrosx 

As in @devrosx video could be seen that his sensor has been set to "Usage", I tried that as well. This doesn't make any difference.

One interesting thing, I did not note before, because I tested only with one value that does not change is that it skips a value.

From your script, you should expect:

1625W, followed by 225W, followed by 123W and then the script start all over again.
The result is 1625, empty, 123, empty, 225, empty 1625. So it skips a message
Exactly, what you see in the movie.

This bash sript is similar to what I did yesterday, when I injected every 2 seconds a URL to Domoticz from Node Red

So, sorry it is still there and it doesn't make any difference, whether it is a bash script, a http message, or a mqtt message.
It is toggling.

**Comment by rwaaren** [comment]
Using script

```
#!/usr/bin/env bash

testID=1298
 
sqlite3 -noheader /opt/domoticz/domoticz.db "select * from devicestatus where id = $testID" 

while :
do
        echo "Press [CTRL+C] to stop.."
        wget -q -O /dev/null "http://192.168.192.115:8084/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=1625;0"
		sqlite3 -noheader /opt/domoticz/domoticz.db "select * from devicestatus where id = $testID" 
        sleep 2
        wget -q -O /dev/null "http://192.168.192.115:8084/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=225;0"
        sqlite3 -noheader /opt/domoticz/domoticz.db "select * from devicestatus where id = $testID" 
        sleep 2
        wget -q -O /dev/null "http://192.168.192.115:8084/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=123;0"
        sqlite3 -noheader /opt/domoticz/domoticz.db "select * from devicestatus where id = $testID" 
        sleep 2
done
```

I get this


```
1298|3|00083298|1|issue4736|1|243|29|0|0|12|255|0|1625;0|2021-03-23 12:43:54|865|0.0|1.0|0.0|1.0|||0|0|0||EnergyMeterMode:MQ==|
▒|2021-03-23 12:43:56|865|0.0|1.0|0.0|1.0|||0|0|0||EnergyMeterMode:MQ==|
1298|3|00083298|1|issue4736|1|243|29|0|0|12|255|0|123;0|2021-03-23 12:43:58|865|0.0|1.0|0.0|1.0|||0|0|0||EnergyMeterMode:MQ==|
Press [CTRL+C] to stop..
▒|2021-03-23 12:44:00|865|0.0|1.0|0.0|1.0|||0|0|0||EnergyMeterMode:MQ==|
1298|3|00083298|1|issue4736|1|243|29|0|0|12|255|0|225;0|2021-03-23 12:44:02|865|0.0|1.0|0.0|1.0|||0|0|0||EnergyMeterMode:MQ==|
```
So something funny ends up in the database.. and likely in the device.


Before
![image](https://user-images.githubusercontent.com/24454798/112141632-cf786180-8bd5-11eb-8953-b73e99fb7653.png)

After

![image](https://user-images.githubusercontent.com/24454798/112141768-05b5e100-8bd6-11eb-9b14-4e827668c821.png)

**Comment by rwaaren** [comment]
Using 

`sqlite3 -noheader /opt/domoticz/domoticz.db "select svalue from devicestatus where id = $testID" | hexdump -c | head -1`

produces

```
0000000   1   6   2   5   ;   0  \n
0000000   0  \r 312   s 017 177  \n
0000000   1   2   3   ;   0  \n
Press [CTRL+C] to stop..
0000000   0  \r 312   s 017 177  \n
0000000   2   2   5   ;   0  \n
0000000   0  \r 312   s 017 177  \n
Press [CTRL+C] to stop..
0000000   1   6   2   5   ;   0  \n
0000000   0  \r 312   s 017 177  \n
0000000   1   2   3   ;   0  \n
Press [CTRL+C] to stop..
0000000   0  \r 312   s 017 177  \n
0000000   2   2   5   ;   0  \n
0000000   0  \r 312   s 017 177  \n
```

**Comment by gizmocuz** [comment]
Those are indeed strange values !! Maybe this does not happen in debug mode...

**Comment by mvdklip** [comment]
> @mvdklip , Best to check if the other issue is now solved and is yes close that issue
> 
> Let's keep this issue open for this problem

I will keep this open, but please note there are three issues really:

1. Domoticz crashing as documented in #4733. I believe the crash itself is fixed although I can't really check because I'm using the stable version of Domoticz. Instead I found a workaround which prevents the crash for me.
2. Strange values / toggling as a result of the fix for #4733. I did not report that because I'm not experiencing and (cannot experience) those problems because I'm running stable. Discussing those problems IMO belongs in #4733 or a separate issue.
3. This issue which is about feeding **negative** instant values to a kWh meter set to computed. This has nothing to do with point 1 or 2 but is simply another bug present in (at least) 2020.2.11995.

NOFI but @FireWizard52 unintentionally hijacked this issue thinking that I was describing the toggling problems, **which I was not**.

**Comment by mvdklip** [comment]
Now carry on discussing the toggling problems here if you want, because it does sound quite nasty and is messing up databases. I will just observe and try to fix the situation with the mixup of two separate issues later. :-D

**Comment by rwaaren** [comment]
I activated debug logging and added some extra debuglines in SQLHelper.cpp

I noticed that for every update from the test bash script the code of line 4873-4880 (see below) is executed twice for the same device id. 
btw. 1298 is the device idx of my test device used in the test script.  

```
_log.Debug(DEBUG_NORM, "SQLH UpdateValueInt line 4873: id = %d ", ulID);
			result = safe_query(
				"UPDATE DeviceStatus SET SignalLevel=%d, BatteryLevel=%d, nValue=%d, sValue='%q', LastUpdate='%04d-%02d-%02d %02d:%02d:%02d' "
				"WHERE (ID = %" PRIu64 ")",
				signallevel, batterylevel,
				nValue, sValue,
				ltime.tm_year + 1900, ltime.tm_mon + 1, ltime.tm_mday, ltime.tm_hour, ltime.tm_min, ltime.tm_sec,
				ulID);

```

```
2021-03-23 22:00:08.297  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:10.315  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:10.414  Debug: SQLH UpdateValueInt issue4736 HwID:3  DevID:00083298 Type:243  sType:29 nValue:0 sValue:123;0
2021-03-23 22:00:12.434  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:14.453  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:14.453  Debug: SQLH UpdateValueInt issue4736 HwID:3  DevID:00083298 Type:243  sType:29 nValue:0 sValue:225;0
2021-03-23 22:00:16.471  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:18.489  Debug: SQLH UpdateValueInt line 4873: id = 1298
2021-03-23 22:00:18.489  Debug: SQLH UpdateValueInt issue4736 HwID:3  DevID:00083298 Type:243  sType:29 nValue:0 sValue:1625;0
2021-03-23 22:00:20.778  Debug: SQLH UpdateValueInt line 4873: id = 1298

```

**Comment by lwalbert** [comment]
Hi @gizmocuz 

Maybe I have solved the toggling issue, atleast it looks OK on my system now.
from line 4740 just below localtime_r(&now, &ltime); replace the code in SQLHelper.cpp
its related to the #4733
```
//Update to fix toggling of object and domoticz crashing
                std::string wpart = result[0][5];
                std::vector<std::string> parts;
                StringSplit(sValue, ";", parts);
                StringSplit(wpart, ";", parts);
                //Commit: If Option 1: energy is computed as usage*time
                //Default is option 0, read from device
                if (options["EnergyMeterMode"] == "1" && (parts.size() == 2) && devType == pTypeGeneral && subType == sTypeKwh)
                {
                        struct tm ntime;
                        double interval;
                        char sCompValue[100];
                        std::string sLastUpdate = result[0][6];
                        time_t lutime;
                        ParseSQLdatetime(lutime, ntime, sLastUpdate, ltime.tm_isdst);

                        interval = difftime(now, lutime);

                                double nEnergy = std::stof(parts[0]) * interval / 3600 + std::stof(parts[1]);
                                sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
                                old_sValue = sCompValue;
                }
               
```
br
lwalbert

**Comment by rwaaren** [comment]
@lwalbert, 

👍 your patch works OK on my system. Tested with the @gizmocuz bash testscript.

**Comment by lwalbert** [comment]
hi @rwaaren 
Thats good to hear, Im still checking the calculation part, so there might be some adjustments.

**Comment by gizmocuz** [comment]
@lwalbert , some questions remarks:

'wpart/parts' is only used when
- options["EnergyMeterMode"] == "1"
- devType == pTypeGeneral && subType == sTypeKwh

So this could be inside the 'if' check, do not check yet for **parts.size() == 2** because there could be faulty data
instead, add a second 'if' below the previous one to check if the parts equals 2
and use the below code between the two 'if' s

                std::vector<std::string> parts;
                StringSplit(sValue, ";", parts);
                StringSplit(wpart, ";", parts);

You use twice 'SplitString' , I thing this one is not needed

                StringSplit(sValue, ";", parts);

I will do some testing as well

**Comment by gizmocuz** [comment]
I think this was the fix:

old_sValue = sCompValue;

**Comment by gizmocuz** [comment]
I'm not sure if the kWh counter is increasing with the above patched code, you are storing the result in 'old_sValue', but that is not used anymore so it does not increment..
Only the Power value is changing in the widget

**Comment by gizmocuz** [comment]
This is the original code from weeks ago:

		if (options["EnergyMeterMode"] == "1" && devType == pTypeGeneral && subType == sTypeKwh)
		{
			std::vector<std::string> parts;
			struct tm ntime;
			double interval;
			float nEnergy;
			char sCompValue[100];
			std::string sLastUpdate = result[0][6];
			time_t lutime;
			ParseSQLdatetime(lutime, ntime, sLastUpdate, ltime.tm_isdst);

			interval = difftime(now, lutime);
			StringSplit(result[0][5], ";", parts);
			nEnergy = static_cast<float>(strtof(parts[0].c_str(), nullptr) * interval / 3600
							 + strtof(parts[1].c_str(), nullptr)); // Rob: whats happening here... strtof ?
			StringSplit(sValue, ";", parts);
			sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
			sValue = sCompValue;
		}

And this caused a crash if sValue did not consisted of two values (separated by a semicolon) 

I made the protection patch again and this is the result:

		if (options["EnergyMeterMode"] == "1" && devType == pTypeGeneral && subType == sTypeKwh)
		{
			std::vector<std::string> parts;
			struct tm ntime;
			double interval;
			float nEnergy;
			char sCompValue[100];
			std::string sLastUpdate = result[0][6];
			time_t lutime;
			ParseSQLdatetime(lutime, ntime, sLastUpdate, ltime.tm_isdst);

			interval = difftime(now, lutime);
			StringSplit(result[0][5], ";", parts);
			if (parts.size() == 2)
			{
				nEnergy = static_cast<float>(std::stof(parts[0]) * interval / 3600 + std::stof(parts[1]));
				StringSplit(sValue, ";", parts);
				if (!parts.empty())
				{
					sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
					sValue = sCompValue;
			
				}
			}
		}

I think it's correct ... should be correct

It is available in beta 13111 ...

**Comment by FireWizard52** [comment]
Unfortunately the problem has not been solved. Energy read has been set to "Computed" .
Upgraded to version 13111 and tested with the previously provided Bash script.
It still toggles
The strange thing is that in addition also the setting (Usage/Return) of another kWh sensor is changed to an "empty" value.
That happens quite often. This setting is not stable.

**Comment by devrosx** [comment]
> Unfortunately the problem has not been solved. Energy read has been set to "Computed" .
> Upgraded to version 13111 and tested with the previously provided Bash script.
> It still toggles
> The strange thing is that in addition also the setting (Usage/Return) of another kWh sensor is changed to an "empty" value.
> That happens quite often. This setting is not stable.

yes something is still wrong...  in gui is still togling and computing not working

**Comment by lwalbert** [comment]
Hi,
The toggling seems to appair only when writing to the sValue.
Compute doesent work probably because the energy value is stil read from device and replaces the computed.
I´we been trying to find a way to separate/deattach the device actual power (W) and energy (Wh) when compute selected.

Managed to get the compute part running yesterday, but then the power (W) from device got stuck/freezed.
I was then writing sValue = result[0][5].c_str(); at the end of compute part if I remember correct.

When compute selected, only the power (W) should be fetched from device, the energy (Wh) should be fetched from database. and returned to database to accumulate by the formula: 
energy(Wh) = power(W) * interval / 3600 + previous value stored in database
update database = energy(Wh)

I made a simple sketch to show what i was triyng to achieve: 
![image001](https://user-images.githubusercontent.com/64552548/112368298-3764a000-8cdb-11eb-8fa5-807d174099c6.png)

Hope fully some one gets an bright idea:)

**Comment by FireWizard52** [comment]
I want to make a suggestion.

As far as I know the Electricity sensor has been working perfect (for me at least) for many years. The last version I used, with a functioning Electricity sensor was version 12084. After updating to version 12090/91 the problem started.

@gizmocuz and @lwalbert did a number a code changes, etc. However none of them solved the problem so far.
Sometimes it is better to do a step back and to jump from there.
So I wonder, if it would not be better to return to the latest version before 12088 that worked.

Of course. there is still the issue that @mvdklip tried to solve. But look to that issue again.
Or do I see it wrong?

**Comment by lwalbert** [comment]
Hi @gizmocuz 

Finally i got it running with compute value, sensor reading and no toggling.
Can  you please test?
```
if (options["EnergyMeterMode"] == "1" && devType == pTypeGeneral && subType == sTypeKwh)
                {
                        std::vector<std::string> parts;
                        struct tm ntime;
                        double interval;
                        float nEnergy;
                        char sCompValue[100];
                        std::string sLastUpdate = result[0][6];
                        time_t lutime;
                        ParseSQLdatetime(lutime, ntime, sLastUpdate, ltime.tm_isdst);

                        interval = difftime(now, lutime);
                        StringSplit(result[0][5], ";", parts);
                        if (parts.size() == 2)
                        {
                                nEnergy = static_cast<float>(std::stof(parts[0]) * interval / 3600 + std::stof(parts[1]));
                                StringSplit(sValue, ";", parts);
                                sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
                                        old_sValue = sCompValue; //result[0][5].c_str();
                        _log.Log(LOG_ERROR, "vaule to db %s", sCompValue);
                        }
                sValue = old_sValue.c_str();
                }
```

**Comment by lwalbert** [comment]
Hi again,
I also believe this resolves the main issue in this case with incorrect negative power value.

positive power 5000W simulated and 8-10W from real meassurement(MQTT):
https://user-images.githubusercontent.com/64552548/112393688-c681b000-8cfb-11eb-9fc0-34ca0e22442d.mov

reverse power 5000W simulated and 8-10W from real meassurement(MQTT):

https://user-images.githubusercontent.com/64552548/112393960-2e37fb00-8cfc-11eb-84b5-8aecc3207ac5.mov

**Comment by mvdklip** [comment]
Nice. Hope this fixes the toggling.

About the negative wattage testing: above video doesn't represent the problem case. Problems appear as soon as the 'Today' value becomes negative. In above video you see a positive value counting down. That has never been a problem. In 2020.2.11995 it stops counting down properly as soon as the Today value reaches 0. From that point on the Today becomes negative (correctly!) but stops counting down properly afterwards. Instead the display value seems to jump back to 0 regularly. I believe this is **not caused by the code in SQLHelper.cpp** because correctly calculated total counters are visible in the log. See first post. Log displays -1278 (Wh) but meter displays -0.080 (kWh). Instead I believe this is a **display and not a calculation problem**. That's why I created a separate issue.

Whether this negative kWh display problem is present in the latest beta is actually unknown. Nobody has tried AFAIK.

**Comment by mvdklip** [comment]
Just updated some of the comments above. Where I was referring to 2020.2 I meant the latest stable release (build 11995).

**Comment by mvdklip** [comment]
> Whether this negative kWh display problem is present in the latest beta is actually unknown. Nobody has tried AFAIK.

Just tested this against builds 11995 and 13111 side by side using the following DzVents script:

	return {
		on = {
			timer = {
				'every minute',
			}
		},
		logging = {
			level = domoticz.LOG_DEBUG,
		},
		execute = function(domoticz, timer)
			domoticz.devices('Negative wattage').updateElectricity(-3600, 0)
			domoticz.devices('Positive wattage').updateElectricity(3600, 0)
		end
	}

Result:

![negative_wattage_13111](https://user-images.githubusercontent.com/200318/112454903-7802fe80-8d59-11eb-99a5-96ca6fee509e.png)

**So the 'negative kWh' problem is not present in build 13111. Only in the latest stable build 11995.**

**Comment by lwalbert** [comment]
Hi @mvdklip 

Oki, retested with negative readings:

https://user-images.githubusercontent.com/64552548/112455059-a5e84300-8d59-11eb-8c4d-02703c3ce867.mov

Looks ok here, I'll keep it running for a while to see.

Ignore the time stamp, had to fool the system since test system has been offline for 10hour.

br
lwalbert

**Comment by mvdklip** [comment]
@lwalbert our messages crossed but I think we reached the same conclusion. The negative kWh problem is not present in the current beta builds. That's nice. Now let's hope the toggling problem has also been solved and can be incorporated in future builds.

**Comment by mvdklip** [comment]
@gizmocuz please have a look at the code provided by lwalbert and see if it can be incorporated. I still don't understand why your code introduced the toggling problem in the first place, but I'm also not sorry I want to understand. ;-)

**Comment by rwaaren** [comment]
I can confirm no toggling occurs with latest @lwalbert patch

**Comment by gizmocuz** [comment]
My test bash script had only send positive values, and there I found no issues.
I never tested with negative values at all... 
I will of course test the patch from @lwalbert , but first I want to see what is wrong
The patch above again does not calculate anymore

Now I see where it goes wrong... 

**if (parts.size() == 2)** (~ line 4755)

This seems NOT to be always the case, how is that possible ?????
the sValue needs to consist of POWER;COUNTER (in this case only POWER and 0, like **1000;0**) and not **1000** only
So somehow someone sends a wrong sValue
This should not happen (never happen)

So what is causing this ? A plugin ? A faulty script ?

So again, test with my bash script above with a dummy general kWh sensor

I created two scripts, one with positive POWER and one with negative.
When I run the positive script for some time I see the today value increasing.
When I run the negative script I see the today value decreasing

![image](https://user-images.githubusercontent.com/2265299/112460528-4b51e580-8d5f-11eb-80e3-8ab15580d54a.png)

@FireWizard52 , It is NOT a good idea to go back to code that causes Domoticz to crash when a invalid sValue was received.
You can always use the stable version, but these safety checks are perfectly valid, just have a look at the code

I made another patch that will alert you in the error log is a invalid sValue is found

		if (options["EnergyMeterMode"] == "1" && devType == pTypeGeneral && subType == sTypeKwh)
		{
			std::vector<std::string> parts;
			struct tm ntime;
			double interval;
			float nEnergy;
			char sCompValue[100];
			std::string sLastUpdate = result[0][6];
			time_t lutime;
			ParseSQLdatetime(lutime, ntime, sLastUpdate, ltime.tm_isdst);

			interval = difftime(now, lutime);
			StringSplit(result[0][5], ";", parts);
			if (parts.size() == 2)
			{
				nEnergy = static_cast<float>(std::stof(parts[0]) * interval / 3600 + std::stof(parts[1]));
				StringSplit(sValue, ";", parts);
				if (!parts.empty())
				{
					sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
					sValue = sCompValue;
			
				}
			}
			else
			{
				//This should not happen!
				_log.Log(LOG_ERROR, "CSQLHelper::UpdateValueInt: Invalid sValue for device idx: %s", result[0][0].c_str());
			}
		}

Here a modified script with a dynamic testID at the top thanks to @rwaaren 

```
#!/usr/bin/env bash

testID=1449

while :
do
        echo "Press [CTRL+C] to stop.."
        wget -q -O /dev/null "http://172.16.0.155:8080/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=1625;0"
        sleep 2
        wget -q -O /dev/null "http://172.16.0.155:8080/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=225;0"
        sleep 2
        wget -q -O /dev/null "http://172.16.0.155:8080/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=123;0"
        sleep 2
done

```

**Comment by devrosx** [comment]
with latest changes... my log 
`2021-03-25 11:59:28.534 MQTT: Topic: domoticz/in, Message: {"idx":249,"command":"udevice","svalue":"164;0"}
2021-03-25 11:59:28.006 Error: CSQLHelper::UpdateValueInt: Invalid sValue for device idx: 940`
these values worked fine...

**Comment by gizmocuz** [comment]
@devrosx , could you please test with my script and a dummy general/kWh sensor ?
(If that works, I will continue testing with MQTT)

**Comment by FireWizard52** [comment]
@gizmocuz 

I agree your argument and I'm in full support of the mentioned safety checks. What I meant to say, that it is sometimes better to do a step back  and to build from there, instead of trying to add more and more.

I tested 13113 on my test system and the toggling is still present, if I configure "Computed".
I used the script that recently has been provided.
But I don't think that the above mentioned changes are incorporated,

**Comment by gizmocuz** [comment]
@FireWizard52 , I stepped 100% back, used the original code, and re-added the safety checks... 
(I assume you created a new dummy general/kWh device and not using the test script on your actual sensor)

**Comment by devrosx** [comment]
> @devrosx , could you please test with my script and a dummy general/kWh sensor ?
> (If that works, I will continue testing with MQTT)

Your scripts works fine on dummy device... (even when i removed nvalue value=0) so payload looks almost same as my MQQT values...

UPDATE
**ooh sorry its not working... (i used Energy read From device on dummy sensor and thats working)
when i set computed Energy read its still toggling..**

**Comment by FireWizard52** [comment]
@gizmocuz 

I was not aware of it. So at the end we had the same thoughts :)

> I assume you created a new dummy general/kWh device

Yes I did, create a complete new device., but still see the toggling

**Comment by FireWizard52** [comment]
@gizmocuz 

I tested it both with JSON over http, as with MQTT from Node Red.
It doesn't matter which of the two I use.

Also the setting of the Type field goes still to empty. This looks random, I tried to track the pattern, but did not succeed so far.

**Comment by mangoose002** [comment]
Hi all,

I have seen that you added a specific error message for this.
Here are some of my results:

2021-03-25 12:35:01.362 MQTT: Topic: domoticz/in, Message: {"idx":1703,"nvalue":0,"svalue":"1.146;0"}
2021-03-25 12:35:01.485 MQTT: Topic: domoticz/in, Message: {"idx":1706,"nvalue":0,"svalue":"0;0"}
2021-03-25 12:35:01.606 MQTT: Topic: domoticz/in, Message: {"idx":1702,"nvalue":0,"svalue":"0;0"}
2021-03-25 12:35:01.363 Error: CSQLHelper::UpdateValueInt: Invalid sValue for device idx: 1703
2021-03-25 12:35:01.487 Error: CSQLHelper::UpdateValueInt: Invalid sValue for device idx: 1706
2021-03-25 12:35:03.330 MQTT: Topic: domoticz/in, Message: {"idx":1704,"nvalue":0,"svalue":"0;0"}
2021-03-25 12:35:03.332 Error: CSQLHelper::UpdateValueInt: Invalid sValue for device idx: 1704 

It does not seem to be consistant idx 1702 & 1704 have the same svalue  and one is OK and not the other.

I hope this can help to solve this.

Regards

**Comment by lwalbert** [comment]
Hi,

My test device is a Zigbee with latest version of stas-demydiuk/domoticz-zigbee2mqtt-plugin running the patch I made based on 13111 running in reverse and no issues. computes and stores the value in database table:Meter
Video of the power meter:
https://user-images.githubusercontent.com/64552548/112468103-2e6de000-8d68-11eb-9531-e5b40e9c7517.mov

<img width="561" alt="Skjermbilde" src="https://user-images.githubusercontent.com/64552548/112469636-fd8eaa80-8d69-11eb-951a-0ba2c6352d1f.png">


I have noticed that occasionally when incorporating a new kwh meter it adds a non working "empty" object where you can't select energy used or return nor from device or computed.
and checked the database, no trace of it in Meter table.
To recover from this state I had to delete the complete plugin folder and clone it again.
So here I suspect some inconsistency in the plugin itselves. But it can also be related to the plugin handler part of domoticz.

Werry strange things going on, I mentioned this in reported issue here:
https://github.com/stas-demydiuk/domoticz-zigbee2mqtt-plugin/issues/605#issue-836733297

**Comment by gizmocuz** [comment]
AHHH okey ! Found (my/the) issue... In debug mode AND release mode (Visual Studio 2019), no issue at all ! (Windows)
But in release mode on a ARM system I see the toggling as well !

Found the issue... : sValue is no **std::string** but a **char***  

Now that's fixed and working with a 'new' general/kWh sensor, but the old sensor I could not change to return/computed anymore
Changing the mode works and it's updated in the database correctly, but because the sValue was incorrect in the database, some fields

Should all be solved now in beta 13115

Thanks for you patience !

**Comment by mangoose002** [comment]
I confirm that I do not have this error anymore in the logs.

**Comment by FireWizard52** [comment]
@gizmocuz 

I tested with the new version 13115 with your earlier provided script, a http JSON call from Node Red and a MQTT message from Node Red.
All sensors behave, what you should expect.  And no errors in the log file.
I will let it run during a couple of hours, but do not expect more issues.

Thanks!!!!

**Comment by gizmocuz** [comment]
Great!!! Thanks for the feedback.

**Comment by devrosx** [comment]
thanks all looks ok for usage only noticed that my return sensor is not updated... (counting is not working)
weird is that i have two return sensors with same setting and one is working...
![Screenshot 2021-03-25 at 14 43 25](https://user-images.githubusercontent.com/13789041/112483116-05a21680-8d79-11eb-9556-d4425c3b7129.jpg)
![Screenshot 2021-03-25 at 14 46 46](https://user-images.githubusercontent.com/13789041/112483121-06d34380-8d79-11eb-94c8-266c50019fc8.jpg)
![Screenshot 2021-03-25 at 14 43 35](https://user-images.githubusercontent.com/13789041/112483124-06d34380-8d79-11eb-8b70-afeed470a96e.jpg)

can anyone test return type

**Comment by lwalbert** [comment]
Hi @gizmocuz 

Just a small change ;-)
Please move the sValue = old_sValue.c_str(); up to line 4765.
The energy read from device doesn't work when inserted after IF at line 4567

```
                 sprintf(sCompValue, "%s;%.1f", parts[0].c_str(), nEnergy);
                                        old_sValue = sCompValue;
                                }
                        }
                        sValue = old_sValue.c_str();
                }
                //sValue = old_sValue.c_str(); ** moved to line 4765. when inserted here, the energy read from device doesent work any more!!
                //~ use different update queries based on the device type
```

Then it all should be solved:)

br
lwalbert

**Comment by gizmocuz** [comment]
Done

**Comment by FireWizard52** [comment]
@gizmocuz,

I'm sorry to say that I came this evening accross one more issue.
It came to my attention, because I was involved in a forum issue.

https://www.domoticz.com/forum/viewtopic.php?f=28&t=22368&p=271855#p271855

This used to work before, but if the following MQTT data is sent and the device has been set to Computed, { command: "udevice", idx: 232, nvalue: 0, svalue: ";0" } Domoticz crashed. This happened in this case, because, when it was getting dark, solar panels stopped producing. This data is empty. So it doesn't accept an empty value.

The strange thing was, that when I changed that to zero "0;0" it also crashed, while my production system (different system, but also sending "0;0" had no problems.

Some additional safety check needed?

**Comment by rwaaren** [comment]
@FireWizard52 
I could reproduce the described behavior when entering ";0" as svalue but "0;0" can be send without issues on my system.
Maybe the first (empty one) left the record in table devicestatus in a faulty state in your test? 

**Open question**: Is this something domoticz should handle gracefully like "garbage in / nothing out"? Or is it the responsibility of the input provider to validate the data send to domoticz?
 

build 13117 on Intel NUC Debian

`mosquitto_pub -h 192.168.192.115 -t 'domoticz/in' -m '{command: "udevice", idx":1298 ,"nvalue":0,"svalue":";0"}'`
`2021-03-26 06:06:28.650  MQTT: Topic: domoticz/in, Message: {"command": "udevice", "idx":1298 ,"nvalue":0,"svalue":";0"}
`
`mosquitto_pub -h 192.168.192.115 -t 'domoticz/in' -m '{command: "udevice", idx":1298 ,"nvalue":0,"svalue":";0"}'`
```
2021-03-26 06:07:34.473  MQTT: Topic: domoticz/in, Message: {"command": "udevice", "idx":1298 ,"nvalue":0,"svalue":";0"}
2021-03-26 06:07:34.474  Error: mosquittodz: Error: stof
```

Tried to disable MQTT module on hardware tab
```
2021-03-26 06:19:31.575  Error: WebServer PO exception occurred : 'stof'
2021-03-26 06:19:33.591  Error: WebServer PO exception occurred : 'stof'
2021-03-26 06:19:35.604  Error: WebServer PO exception occurred : 'stof'
....
2021-03-26 06:19:57.594  Error: WebServer PO exception occurred : 'stof'
2021-03-26 06:22:22.999  Error: WebServer:8084 thread seems to have ended unexpectedly (last update 73.000000 seconds ago)
```


```
service domoticz stop
cp backups/hourly/backup-hour-06-domoticz.db  domoticz.db
rm *shm
rm *wal
service domoticz start
```

`mosquitto_pub -h 192.168.192.115 -t 'domoticz/in' -m '{"command": "udevice", "idx":1298 ,"nvalue":1,"svalue":"0;0"}'`
```

2021-03-26 06:41:35.959  MQTT: Topic: domoticz/in, Message: {"command": "udevice", "idx":1298 ,"nvalue":1,"svalue":"0;0"}
2021-03-26 06:41:51.108  MQTT: Topic: domoticz/in, Message: {"command": "udevice", "idx":1298 ,"nvalue":1,"svalue":"0;0"}
2021-03-26 06:42:31.416  MQTT: Topic: domoticz/in, Message: {"command": "udevice", "idx":1298 ,"nvalue":1,"svalue":"0;0"}
```

No issues with this value for svalue

**Comment by kiddigital** [comment]
@rwaaren 
> 
> **Open question**: Is this something domoticz should handle gracefully like "garbage in / nothing out"? Or is it the responsibility of the input provider to validate the data send to domoticz?
>

In general terms it is considered good (coding) practice that a program validates the input it is getting to see if the input conforms to the expected requirements before processing the input. Makes programs more robust (should not crash) and more secure (less abuse).

Difficulty for a system like Domoticz that it consists of a lot of different parts that have to play together. Example, should it be the (Angular) UI to check the users input? But what about scripts, etc. that do not use the UI?

So it makes more sense that the Domoticz webserver should do the checking, as that is an entry point everyone and everything uses to communicate with Domoticz. But at the webserver 'level' there is no knowledge (yet) of the input (and/or output) that is expected in each of the internal hardware modules. So not easy to check there.

(But check my WIP on adding OpenAPI to the webserver. A step into the right direction I think)

There are also some other intermediate 'layers' like MQTT that play a similar role as the webserver receiving data without exactly knowing what to expect as that is handled downstream by the hardware modules again. Not easy to do there.

And a lot (most?) communication between Domoticz and the outside world is done by the hardware modules themselves. They talk directly to the hardware and are the only ones to exactly know how the input (and output) should behave within there limited scope.

Maybe there is a need for some kind of a 'layer' in the codebase that could be leveraged as a more generic input validation layer a module could use by 'just' declaring the input (and output) specifications without the need the programmatically build all the checking each time.

Could work for a lot of pieces except those hardware modules that are hardwired to directly attached hardware which will always use direct communication.

It would be awesome if input providers take their responsibility and made sure that their output is always valid output, but... unfortunately that will for sure not always be the case leaving no alternative to Domoticz to pick up the burden of doing checks as well 🙁

**Comment by gizmocuz** [comment]
I added a extra safety try/catch in the UpdateDevice function. (beta 13118)
The correct JSON API is explained on our wiki. Plugin developers should comply with them.

**Comment by rwaaren** [comment]
@kiddigital Thanks for your extensive answer / points of view. Much appreciated.

@gizmocuz the extra try/catch does help because webserver does not crash but it does not prevent the svalue from being hampered leading to one line

`2021-03-26 09:20:47.936  Error: MainWorker::UpdateDevice exception occurred : 'stof'`

for every given

`/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=1625;0`

after sending a ";0" (mosquitto) 

Only after svalue is corrected directly editing the database with sqlite new updates are reaching the record.


Not sure if the intention of your commit was to prevent this tough.

**Comment by devrosx** [comment]
updated but something is still wrong... one of my energy return sensor (Computed is not working)
weird is that another one works great (with same setting and data)

this is on json, data is wrong... but, no errors in log (and no updates in history)
im using MQQT to update 
domoticz LOG
2021-03-26 09:50:03.021 MQTT: Topic: domoticz/in, Message: {"idx":249,"command":"udevice","svalue":"106;0"}

`{
	"ActTime" : 1616748435,
	"AstrTwilightEnd" : "20:14",
	"AstrTwilightStart" : "03:57",
	"CivTwilightEnd" : "18:55",
	"CivTwilightStart" : "05:17",
	"DayLength" : "12:34",
	"NautTwilightEnd" : "19:34",
	"NautTwilightStart" : "04:38",
	"ServerTime" : "2021-03-26 09:47:15",
	"SunAtSouth" : "12:06",
	"Sunrise" : "05:49",
	"Sunset" : "18:23",
	"app_version" : "2020.2 (build 13118)",
	"result" : 
	[
		{
			"AddjMulti" : 1.0,
			"AddjMulti2" : 1.0,
			"AddjValue" : 0.0,
			"AddjValue2" : 0.0,
			"BatteryLevel" : 255,
			"CustomImage" : 0,
			"Data" : "0, `\ufffd\ufffd",
			"Description" : "",
			"EnergyMeterMode" : "1",
			"Favorite" : 1,
			"HardwareDisabled" : false,
			"HardwareID" : 14,
			"HardwareName" : "NET",
			"HardwareType" : "Dummy (Does nothing, use for virtual switches only)",
			"HardwareTypeVal" : 15,
			"ID" : "00082249",
			"LastUpdate" : "2021-03-26 09:46:44",
			"Name" : "Sol\u00e1ry - celkem",
			"Notifications" : "false",
			"PlanID" : "7",
			"PlanIDs" : 
			[
				7,
				10
			],
			"Protected" : false,
			"ShowNotifications" : true,
			"SignalLevel" : "-",
			"SubType" : "kWh",
			"SwitchTypeVal" : 4,
			"Timers" : "false",
			"Type" : "General",
			"TypeImg" : "current",
			"Unit" : 1,
			"Used" : 1,
			"XOffset" : "0",
			"YOffset" : "0",
			"idx" : "249"
		}
	],
	"status" : "OK",
	"title" : "Devices"
}`

**Comment by kiddigital** [comment]
> 
> @kiddigital Thanks for your extensive answer / points of view. Much appreciated.

yw :)

> 
> @gizmocuz the extra try/catch does help because webserver does not crash but it does not prevent the svalue from being hampered leading to one line
> 
> `2021-03-26 09:20:47.936 Error: MainWorker::UpdateDevice exception occurred : 'stof'`
> 
> for every given
> 
> `/json.htm?type=command&param=udevice&idx=$testID&nvalue=0&svalue=1625;0`
> 
> after sending a ";0" (mosquitto)
> 
> Only after svalue is corrected directly editing the database with sqlite new updates are reaching the record.
> 
> Not sure if the intention of your commit was to prevent this tough.

What does sValue contain in the database before you 'fix' it? Even when the current DB value is wrong, you would expect updating with a correctly received sValue would overwrite the wrong value. But it seems that the SQL update fails because of the wrong value? Or is there a 'read' operation in front of the update that fails because of the wrong value which causes the update to never take place? (sounds logical, haven't debugged the code).

Should we (re)open the/a (new) issue for this?

**Comment by gizmocuz** [comment]
Feels like cleaning up someone else code and plugin issues here...
So this time I used the old school atof functions again... beta 13119

@kiddigital , this is a special sensor that creates it's own sValue, implementation code was not very clean (not pointing fingers to anyone)
If the value in the database was already wrong (like ";123" instead of "123;456") it was not altered because of a exception caused by calling std::stof on a empty string

**Comment by devrosx** [comment]
So my data in sensor value in json ("Data" : "0, `\ufffd\ufffd",)
is because previous value in database was wrong (because new data are OK)?
any idea how to fix it?

**Comment by gizmocuz** [comment]
@devrosx , did you update to the latest beta, did that fix it ?
If not, edit the field in the database, or delete the sensor and it will be added back

**Comment by devrosx** [comment]
just updated and still same
data on device "Data" : "0, `\ufffd\ufffd",

**Comment by gizmocuz** [comment]
Okey, so try to correct it in your database (make a backup from the settings, stop domoticz, edit the database, start domoticz, restore from settings)
Or delete the sensor
If it is still wrong after delete/recreate then have a look at the process/plugin that update data

**Comment by rwaaren** [comment]
Try 
```
wget -q -O /dev/null "http://<domoticz ip: domoticz port>/json.htm?type=command&param=udevice&idx=<device id>&nvalue=0&svalue=0;0"
```

**Comment by gizmocuz** [comment]
(or just put in another browser window)
`http://<domoticz ip: domoticz port>/json.htm?type=command&param=udevice&idx=<device id>&nvalue=0&svalue=0;0`

**Comment by devrosx** [comment]
Thanks.... finally it works

what i did:
switch sensor **from device** instead of **computed**
then send
`http://<domoticz ip: domoticz port>/json.htm?type=command&param=udevice&idx=<device id>&nvalue=0&svalue=0;0
`
now right value is saved and all works (switched back to computed)

when i send 0;0 value without switch sensor it wasn't working ....

**Comment by FireWizard52** [comment]
Thanks for fixing this issue.
I tested it with all kind of values, but it doesn't crash . Great!

To answer @rwaaren question.
Should we fix these kind of problems in Domoticz or is it the responsibility of the input provider to validate the data send to Domoticz?.

I think both. Both parties has the responsibility to prevent possible crashes.
I fully agree with @kiddigital and his extensive explanation.

One problem we meet is that not always every information is available. I became involved In the project with the APSystems inverters approx. 1.5 year ago. Other are going to use a application, solution. flow or script as well and you don know what their input is. As an example, this empty string value used to work until Domoticz was modified (this week). It is improved and sooner or later this code will also been used in a future stable release.  In this case  a lot of users will experience crashes after they updated a working installation. So Domoticz should check and this avoids that the forum explodes in case of an update,

But once again thanks for the relative quick fix. This would never had happen with proprietary software.

Regards

**Comment by kiddigital** [comment]
> 
> 
> Thanks for fixing this issue.
> I tested it with all kind of values, but it doesn't crash . Great!
> 
>....
> 
> But once again thanks for the relative quick fix. This would never had happen with proprietary software.
> 
> Regards

What would never have happened... The problem or the quick fix ? ;)

**Comment by FireWizard52** [comment]
The quick fix :)
